src/logic/mul.rs

```diff
@@ -0,0 +1,403 @@
+                            }
+                            result.wrapping_lshr_assign(8usize).unwrap();
+                            assert_eq!(result.clone().resize_to_u8(),u8::MAX);
+                        }
+                    }
+                }
+            }
+            //test inl apints
+            assert_eq!(
+                ApInt::from(u8::MAX)
+                    .into_wrapping_mul(&ApInt::from(u8::MAX))
+                    .unwrap(),
+                ApInt::from(1u8)
+            );
+            nine_test(2);
+            nine_test(3);
+            nine_test(4);
+            nine_test(7);
+            nine_test(8);
+            //test ext apints
+            nine_test(9);
+            nine_test(16);
+            //5 digits wide
+            nine_test(40);
+            nine_test(63);
+            //non overflowing test
+            let resize = [
+                7usize, 8, 9, 15, 16, 17, 31, 32, 33, 63, 64, 65, 127, 128, 129, 137, 200, 255,
+                256, 700, 907, 1024, 2018, 2019,
+            ];
+            let lhs_shl = [
+                0usize, 1, 0, 1, 4, 7, 4, 10, 13, 0, 31, 25, 7, 17, 32, 50, 0, 64, 249, 8, 777, 0,
+                1000, 0,
+            ];
+            let rhs_shl = [
+                0usize, 0, 1, 1, 3, 6, 4, 14, 10, 0, 0, 25, 0, 18, 32, 49, 100, 64, 0, 256, 64,
+                900, 1000, 0,
+            ];
+            for (i, _) in resize.iter().enumerate() {
+                let lhs = ApInt::from(5u8)
+                    .into_zero_resize(BitWidth::new(resize[i]).unwrap())
+                    .into_wrapping_shl(lhs_shl[i])
+                    .unwrap();
+                let rhs = ApInt::from(11u8)
+                    .into_zero_resize(BitWidth::new(resize[i]).unwrap())
+                    .into_wrapping_shl(rhs_shl[i])
+                    .unwrap();
+                let zero = ApInt::from(0u8).into_zero_resize(BitWidth::new(resize[i]).unwrap());
+                let one = ApInt::from(1u8).into_zero_resize(BitWidth::new(resize[i]).unwrap());
+                let expected = ApInt::from(55u8)
+                    .into_zero_resize(BitWidth::new(resize[i]).unwrap())
+                    .into_wrapping_shl(rhs_shl[i] + lhs_shl[i])
+                    .unwrap();
+                assert_eq!(lhs.clone().into_wrapping_mul(&zero).unwrap(), zero);
+                assert_eq!(zero.clone().into_wrapping_mul(&rhs).unwrap(), zero);
+                assert_eq!(lhs.clone().into_wrapping_mul(&one).unwrap(), lhs);
+                assert_eq!(one.clone().into_wrapping_mul(&rhs).unwrap(), rhs);
+                assert_eq!(lhs.clone().into_wrapping_mul(&rhs).unwrap(), expected);
+            }
+            assert_eq!(
+                ApInt::from([0,0,0,0,u64::MAX,0,u64::MAX,u64::MAX])
+                .into_wrapping_mul(&ApInt::from([0,0,0,0,u64::MAX,u64::MAX,0,u64::MAX])).unwrap()
+                ,ApInt::from([u64::MAX,0,1,u64::MAX - 3,1,u64::MAX,u64::MAX,1]));
+        }
+    }
+}
```

Comment:
As for the division algorithms I won't take ownership or maintenance for this algorithm.
Please confirm that you are going to overtake maintenance for this piece of code in the future before I merge it.

Reply:
I will overtake maintenance 100%. In worst case scenarios, I can quickly implement a binary division routine, but hopefully I can simply continue forward.